rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py

```diff
@@ -341,72 +363,255 @@ def _convert_attribute_tokens_to_texts(
-                "Leaving an untrained CountVectorizer"
+                f"New data contains vocabulary of size {len(new_vocabulary)} for attribute {attribute} "
+                f"which is larger than the maximum vocabulary size({len(existing_vocabulary)}) "
+                f"of the original model. Some tokens will have to be dropped "
+                f"in order to continue training. It is advised to re-train the "
+                f"model from scratch on the complete data."
             )
+        self._merge_new_vocabulary_tokens(existing_vocabulary, new_vocabulary)
+        self._set_vocabulary(attribute, existing_vocabulary)
+
+    def _merge_new_vocabulary_tokens(
+        self, existing_vocabulary: Dict[Text, int], vocabulary: Set[Text]
+    ):
+        available_empty_index = self._get_starting_empty_index(existing_vocabulary)
+        for token in vocabulary:
+            if token not in existing_vocabulary:
+                existing_vocabulary[token] = available_empty_index
+                del existing_vocabulary[f"{BUFFER_SLOTS_PREFIX}{available_empty_index}"]
+                available_empty_index += 1
+                if available_empty_index == len(existing_vocabulary):
+                    # We have exhausted all available vocabulary slots.
+                    # Drop the remaining vocabulary.
+                    break
+
+    def _get_additional_vocabulary_size(
+        self, attribute: Text, existing_vocabulary_size: int
+    ) -> int:
+        """Gets additional vocabulary size to be saved for incremental training.
+
+        If `self.additional_vocabulary_size` is not `None`,
+        we return that as the user should have specified
+        this number. If not then we take the default
+        additional vocabulary size which is 1/2 of the
+        current vocabulary size.
+
+        Args:
+            attribute: Message attribute for which additional vocabulary size should be computed.
+            existing_vocabulary_size: Current size of vocabulary learnt from the training data.
+
+        Returns:
+            Size of additional vocabulary that should be set aside for incremental training.
+        """
+        # Vocabulary expansion for INTENTS, ACTION_NAME
+        # and INTENT_RESPONSE_KEY is currently not supported as
+        # incremental training does not support creation/deletion
+        # of new/existing labels(intents, actions, etc.)
+        if attribute not in DENSE_FEATURIZABLE_ATTRIBUTES:
+            return 0
+        if self.additional_vocabulary_size.get(attribute) is not None:
+            return self.additional_vocabulary_size[attribute]
+
+        # If the user hasn't defined additional vocabulary size,
+        # then we increase it by 1000 minimum. If the current
+        # vocabulary size is greater than 2000, we take half of
+        # that number as additional vocabulary size.
+        return max(MIN_ADDITIONAL_CVF_VOCABULARY, int(existing_vocabulary_size * 0.5))
```

Comment:
A bit of both. The maximum number is based on some experiments I ran which showed that adding more than 3/4th of current vocabulary size hurt the performance of training models from scratch. Also, a very large number will slow down the training as well because of larger matrix multiplications. So, 0.5 of existing vocabulary size seemed like a good default. The minimum number is inferred from vocabulary sizes of datasets in our training data repo.